File: cpp/LeetCode/No209.cpp
/*
Given an array of positive integers nums and a positive integer target, return the minimal length of a subarray whose sum is greater than or equal to target.
If there is no such subarray, return 0 instead.
*/
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    int minSubArrayLen(int target, vector<int>& nums) {
        int res = nums.size()+1;
        int s = 0;
        int left = 0, right = 0;
        while(right<nums.size()){
            s+=nums[right];
            while(s>=target){
                res = min(res, right-left+1);
                s-=nums[left++];
            }
            right++;
        }
        return res<nums.size()+1?res:0;
    }
};

int main(void){
    Solution solution;
    int target;
    vector<int> nums;

    /*target = 7;
    nums = {2,3,1,2,4,3};
    cout << solution.minSubArrayLen(target, nums) << endl;

    target = 4;
    nums = {1,4,4};
    cout << solution.minSubArrayLen(target, nums) << endl;

    target = 11;
    nums = {1,1,1,1,1,1,1,1};
    cout << solution.minSubArrayLen(target, nums) << endl;*/

    target = 11;
    nums = {1,2,3,4,5};
    cout << solution.minSubArrayLen(target, nums) << endl;
    return 0;
}

File: python/LeetCode Quiz/No209.py
'''
Given an array of positive integers nums and a positive integer target
, return the minimal length of a contiguous subarray [numsl, numsl+1, ..., numsr-1, numsr]
 of which the sum is greater than or equal to target. 
 
 If there is no such subarray, return 0 instead.
'''
from typing import List

class Solution:
    def minSubArrayLen(self, target: int, nums: List[int]) -> int:
        temp_sum = 0
        left = 0
        res = 100001
        for i, val in enumerate(nums):
            temp_sum+=val
            while temp_sum>target:
                temp_sum-=nums[left]
                res = min(res, i-left+1)
                left+=1
        return res if res<100001 else 0

def main():
    target = 11
    nums = [1,1,3,4,5]
    print(Solution.minSubArrayLen(Solution(), target, nums))

if __name__=='__main__':
    main()
